### PR TITLE
Supports super().__torch_dispatch__ with arguments list

### LINKED ISSUE #73933: Calling super().__torch_dispatch__ with arguments list causes segfault
### 🐛 Describe the bug

https://github.com/albanD/subclass_zoo/blob/33d7afe63c2a336e01eaf3e81fba085a68e3955f/bug_zoo.py#L18-L24

```
        class SuperDispatchSegfaultTensor(BaseTensor):
            @classmethod
            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                return super().__torch_dispatch__(func, types, list(args), kwargs)
        SuperDispatchSegfaultTensor(torch.tensor(1.0)).neg()
```

This segfaults probably because not doing enough type check in CPython. Maybe `__torch_function__` also has same problem.

### Versions

master

cc @albanD @zou3519

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74509

Summary:
For THPModule_disable_torch_(dispatch|function),  converts list arguments to tuple before invoking PyObject_Call.

Fixes  #73933

Differential Revision: [D35034204](https://our.internmc.facebook.com/intern/diff/D35034204)

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

File: torch/csrc/utils/disable_torch_function.cpp [hunk 1]
```diff
@@ -119,8 +119,9 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
   py::tuple py_args;
   if (args == nullptr) {
     py_args = py::make_tuple();
-  }
-  else {
+  } else if (PyList_CheckExact(args)) {
+    py_args = py::reinterpret_borrow<py::tuple>(PyList_AsTuple(args));
```
Comment (ezyang) [inline torch/csrc/utils/disable_torch_function.cpp:123]: reinterpret_borrow here is wrong as PyList_AsTuple returns a new reference; if you only borrow it you will leak the allocated tuple here. https://github.com/pybind/pybind11/issues/1201 may be of interest.

**Review by ezyang — CHANGES_REQUESTED** [review]
Current implementation leaks PyObjects.

File: torch/csrc/utils/disable_torch_function.cpp (hunk 1 above)
Comment (SherlockNoMad) [reply torch/csrc/utils/disable_torch_function.cpp:123]: ah... I need to use reinterpret_steal here, as a move semantics. Thanks a lot for the cheatsheet, it's way clearer than pybind's doc. :)

**Comment by SherlockNoMad** [comment]
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

**Review by ezyang — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

**Pull request closed without merging**